src/Core/Extensible.php

```diff
@@ -549,21 +556,26 @@ public function getExtensionInstances()
             if (in_array($class, self::$unextendable_classes)) {
                 continue;
             }
-            $extensions = Config::inst()->get($class, 'extensions', Config::UNINHERITED | Config::EXCLUDE_EXTRA_SOURCES);
-
-            if ($extensions) {
-                foreach ($extensions as $extension) {
-                    $name = $extension;
-                    // Allow service names of the form "%$ServiceName"
-                    if (substr($name, 0, 2) == '%$') {
-                        $name = substr($name, 2);
-                    }
-                    $name = trim(strtok($name, '('));
-                    if (class_exists($name)) {
-                        $name = ClassInfo::class_name($name);
-                    }
-                    $this->extension_instances[$name] = Injector::inst()->get($extension);
+            $extensions = Config::inst()->get(
+                $class,
+                'extensions',
+                Config::UNINHERITED | Config::EXCLUDE_EXTRA_SOURCES
+            ) ?? [];
+
+            foreach ($extensions as $extension) {
+                // Allow service names of the form "%$ServiceName" or "ServiceName.default"
+                [$extensionClass] = ClassInfo::parse_class_spec($extension);
+
+                // If singletons have constructor dependencies they must be service references, and
+                // must be defined in Injector configuration
```

Comment:
I'm confused what this comment means.  Could you elaborate what this means in your comment, maybe provide an example?

Reply:
Would it make more sense if I said "must be defined in Injector configuration for that singleton"?

Reply:
yes adding those last few words does help